# Download notebook

```default
!wget https://raw.githubusercontent.com/miruetoto/yechan3/main/posts/2_Studies/PyG/ls3.ipynb
```

# 미니배치

In [4]:
import torch_geometric

In [5]:
dataset = torch_geometric.datasets.TUDataset(
    root='/tmp/ENZYMES', 
    name='ENZYMES'
)

> (ChatGPT) ENZYMES는 그래프 분류를 위한 벤치마크 데이터셋 중 하나입니다. 이 데이터셋은 600개의 그래프로 구성되어 있으며, 6개의 클래스로 분류됩니다. 각 그래프는 효소(enzyme) 분자의 구조를 나타내며, 그래프의 노드는 원자(atom)를 나타내고, 엣지(edge)는 원자 간의 연결을 나타냅니다. ENZYMES 데이터셋은 화학 및 생물 정보학 분야에서 그래프 분류 알고리즘의 성능을 평가하기 위해 사용될 수 있습니다. 그래프 분류 알고리즘은 주어진 그래프를 특정 클래스 레이블로 분류하는 작업을 수행하는데 사용됩니다. 예를 들어, ENZYMES 데이터셋의 그래프는 특정 효소 종류를 나타내며, 그래프 분류 알고리즘은 주어진 효소 그래프가 어떤 종류의 효소인지 예측할 수 있습니다. PyG를 사용하여 ENZYMES 데이터셋을 초기화하면 해당 데이터셋을 다운로드하고 필요한 전처리를 자동으로 수행할 수 있습니다. 그래프 데이터를 다루는 머신 러닝 모델을 구축하고 훈련시키기 위해 ENZYMES 데이터셋을 사용할 수 있습니다.

In [6]:
len(dataset) # 이 데이터셋에는 600개의 그래프가 있음

600

In [7]:
dataset.num_classes # 6개의 클래스

6

In [8]:
dataset.num_node_features # 각 노드에는 3개의 피처가 있음

3

`-` 600개의 그래프중 첫번째 그래프에 접근 

In [9]:
dataset[0]

Data(edge_index=[2, 168], x=[37, 3], y=[1])

- `x=[37, 3]`: $|{\cal V}|=37$, $f \in \mathbb{R}^3$
- `edge_index=[2, 168]`: $|{\cal E}|=168$

`-` 600개중에서 두번째 그래프에 접근

In [10]:
dataset[1]

Data(edge_index=[2, 102], x=[23, 3], y=[1])

- `x=[23, 3]`: $|{\cal V}|=23$, $f \in \mathbb{R}^3$
- `edge_index=[2, 102]`: $|{\cal E}|=102$

`-` dataset $\to$ loader 

In [37]:
loader = torch_geometric.loader.DataLoader(dataset, batch_size=2, shuffle=False)

In [38]:
for i,batch in enumerate(loader):
    if i<5:
        print(i,batch)

0 DataBatch(edge_index=[2, 270], x=[60, 3], y=[2], batch=[60], ptr=[3])
1 DataBatch(edge_index=[2, 182], x=[49, 3], y=[2], batch=[49], ptr=[3])
2 DataBatch(edge_index=[2, 182], x=[47, 3], y=[2], batch=[47], ptr=[3])
3 DataBatch(edge_index=[2, 384], x=[114, 3], y=[2], batch=[114], ptr=[3])
4 DataBatch(edge_index=[2, 184], x=[55, 3], y=[2], batch=[55], ptr=[3])


- 600개 그래프를 2개씩 묶어서 배치를 만듬

In [39]:
dataset[0], dataset[1]

(Data(edge_index=[2, 168], x=[37, 3], y=[1]),
 Data(edge_index=[2, 102], x=[23, 3], y=[1]))

 - 이게 합쳐져서 `0 DataBatch(edge_index=[2, 270], x=[60, 3], y=[2], batch=[60], ptr=[3])`

In [41]:
dataset[2], dataset[3]

(Data(edge_index=[2, 92], x=[25, 3], y=[1]),
 Data(edge_index=[2, 90], x=[24, 3], y=[1]))

 - 이게 합쳐져서 `1 DataBatch(edge_index=[2, 182], x=[49, 3], y=[2], batch=[49], ptr=[3])`